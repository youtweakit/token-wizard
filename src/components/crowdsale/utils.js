import { attachToContract } from '../../utils/blockchainHelpers'
import { noContractAlert } from '../../utils/alerts'
import { toFixed } from '../../utils/utils'
import { CONTRACT_TYPES } from '../../utils/constants'
import { contractStore, crowdsalePageStore, tokenStore, web3Store } from '../../stores'
import { toJS } from 'mobx'
import { BigNumber } from 'bignumber.js'

BigNumber.config({ DECIMAL_PLACES : 18 })

export const toBigNumber = (value) => isNaN(value) || value === '' ? new BigNumber(0) : new BigNumber(value)


export function getJoinedTiers(abi, addr, joinedCrowdsales, cb) {
  attachToContract(abi, addr)
    .then(crowdsaleContract => {
      console.log('attach to crowdsale contract')

      crowdsaleContract.methods.joinedCrowdsalesLen().call(function (err, joinedCrowdsalesLen) {
        if (err) {
          console.log(err)
          return cb([])
        }

        getJoinedTiersRecursively(0, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, function (_joinedCrowdsales) {
          cb(_joinedCrowdsales)
        })
      })
    })
    .catch(err => {
      console.log(err)
      cb([])
    })
}

function getJoinedTiersRecursively(i, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, cb) {
  if (i >= joinedCrowdsalesLen) {
    return cb(joinedCrowdsales);
  }

  crowdsaleContract.methods.joinedCrowdsales(i).call(function(err, joinedCrowdsale) {
    if (err) return console.log(err);
    console.log("joinedCrowdsale: " + joinedCrowdsale);

    if (joinedCrowdsale === "0x") {
      cb(joinedCrowdsales);
    } else {
      joinedCrowdsales.push(joinedCrowdsale);
      i++;
      getJoinedTiersRecursively(i, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, cb);
    }
  })
}

export function findCurrentContractRecursively(i, firstCrowdsaleContract, cb) {
  console.log(contractStore.crowdsale.addr);
  let crowdsaleAddr = contractStore.crowdsale.addr[i];
  const { web3 } = web3Store

  if (i === contractStore.crowdsale.addr.length) return cb(firstCrowdsaleContract, i);
  if (!crowdsaleAddr) return cb(null);
  if (!web3.utils.isAddress(crowdsaleAddr)) return cb(null);

  attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
    .then(crowdsaleContract => {
      console.log('attach to crowdsale contract')

      if (i === 0) {
        firstCrowdsaleContract = crowdsaleContract
      }

      if (!crowdsaleContract) return noContractAlert()

      console.log(contractStore.crowdsale.contractType)

      if (contractStore.crowdsale.contractType === CONTRACT_TYPES.standard) return cb(crowdsaleContract, i)

      crowdsaleContract.methods.startsAt().call(function (err, startDate) {
        if (err) return console.log(err)

        startDate = startDate * 1000
        console.log('startDate: ' + startDate)
        crowdsaleContract.methods.endsAt().call(function (err, endDate) {
          if (err) return console.log(err)

          endDate = endDate * 1000
          console.log('endDate: ' + endDate)

          let curDate = new Date().getTime()
          console.log('curDate: ' + curDate)
          if (curDate < endDate && curDate >= startDate) {
            cb(crowdsaleContract, i)
          } else {
            i++
            findCurrentContractRecursively(i, firstCrowdsaleContract, cb)
          }
        })
      })
    })
    .catch(err => console.log(err))
}

export function getCrowdsaleTargetDates($this, cb) {
  let propsCount = 0;
  let cbCount = 0;
  let state = $this.state;
  const { web3 } = web3Store

  for (let i = 0; i < contractStore.crowdsale.addr.length; i++) {
    let crowdsaleAddr = contractStore.crowdsale.addr[i];
    attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
      .then(crowdsaleContract => { // eslint-disable-line no-loop-func
        console.log("attach to crowdsale contract");

        if (!crowdsaleContract) return noContractAlert();

        if (crowdsaleContract.methods.startsAt) {
          propsCount++;
          crowdsaleContract.methods.startsAt().call((err, startDate) => {
            cbCount++;
            if (err) return console.log(err);

            console.log("startDate: " + startDate * 1000);
            if (!crowdsalePageStore.startDate || crowdsalePageStore.startDate > startDate * 1000)
              crowdsalePageStore.startDate = startDate * 1000;
            if (propsCount === cbCount) {
              state.loading = false;
              $this.setState(state, cb);
            }
          });
        }

        if (crowdsaleContract.methods.endsAt) {
          propsCount++;
          crowdsaleContract.methods.endsAt().call((err, endDate) => {
            cbCount++;
            if (err) return console.log(err);

            console.log("endDate: " + endDate * 1000);

            if (!crowdsalePageStore.endDate || crowdsalePageStore.endDate < endDate * 1000) {
              crowdsalePageStore.setProperty('endDate', endDate * 1000)
            }

            console.log("curDate: " + new Date().getTime());

            if (propsCount === cbCount) {
              state.loading = false;
              $this.setState(state, cb);
            }
          });
        }
      })
      .catch(console.log)
  }
}

export function initializeAccumulativeData() {
  crowdsalePageStore.setProperty('maximumSellableTokens', 0)
  crowdsalePageStore.setProperty('maximumSellableTokensInWei', 0)
  crowdsalePageStore.setProperty('investors', 0)
  crowdsalePageStore.setProperty('ethRaised', 0)
  crowdsalePageStore.setProperty('weiRaised', 0)
  crowdsalePageStore.setProperty('tokensSold', 0)
  crowdsalePageStore.setProperty('tokenAmountOf', 0)
  return Promise.resolve()
}

export function getAccumulativeCrowdsaleData() {
  const { web3 } = web3Store

  let promises = contractStore.crowdsale.addr
    .map(crowdsaleAddr => {
      return attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
        .then(crowdsaleContract => { // eslint-disable-line no-loop-func
          if (!crowdsaleContract) return noContractAlert()

          let getWeiRaised = crowdsaleContract.methods.weiRaised().call().then((weiRaised) => {
            let newWeiRaised
            if (crowdsalePageStore.weiRaised) {
              newWeiRaised = crowdsalePageStore.weiRaised + parseInt(weiRaised, 10);
            } else {
              newWeiRaised = parseInt(weiRaised, 10);
            }

            crowdsalePageStore.setProperty('weiRaised', newWeiRaised)
            crowdsalePageStore.setProperty('ethRaised', parseFloat(web3.utils.fromWei(toFixed(crowdsalePageStore.weiRaised).toString(), 'ether')))
          })

          let getTokensSold = crowdsaleContract.methods.tokensSold().call().then((tokensSold) => {
            if (crowdsalePageStore.tokensSold) {
              crowdsalePageStore.setProperty('tokensSold', crowdsalePageStore.tokensSold + parseInt(tokensSold, 10))
            } else {
              crowdsalePageStore.setProperty('tokensSold', parseInt(tokensSold, 10))
            }
          })

          let getMaximumSellableTokens = crowdsaleContract.methods.maximumSellableTokens().call().then((maximumSellableTokens) => {
            const maxSellableTokens = toBigNumber(crowdsalePageStore.maximumSellableTokens)
            crowdsalePageStore.setProperty('maximumSellableTokens', maxSellableTokens.plus(maximumSellableTokens).toFixed())

            //calc maximumSellableTokens in Eth
            return setMaximumSellableTokensInEth(crowdsaleContract, maximumSellableTokens)
          })

          let getInvestors = crowdsaleContract.methods.investorCount().call().then((investors) => {
            const oldInvestors = crowdsalePageStore.investors
            const investorsCount = parseInt(investors, 10)

            if (oldInvestors) {
              crowdsalePageStore.setProperty('investors', oldInvestors + investorsCount);
            } else {
              crowdsalePageStore.setProperty('investors', investorsCount);
            }
          });

          return Promise.all([getWeiRaised, getTokensSold, getMaximumSellableTokens, getInvestors])
        })
    })

  return Promise.all(promises)
}

function setMaximumSellableTokensInEth(crowdsaleContract, maximumSellableTokens) {
  return crowdsaleContract.methods.pricingStrategy().call()
    .then((pricingStrategyAddr) => {
      return attachToContract(contractStore.pricingStrategy.abi, pricingStrategyAddr)
    })
    .then(pricingStrategyContract => {
      if (!pricingStrategyContract) return noContractAlert()

      return pricingStrategyContract.methods.oneTokenInWei().call()
        .then((oneTokenInWei) => {
          const currentMaximumSellableTokensInWei = toBigNumber(crowdsalePageStore.maximumSellableTokensInWei)
          const maximumSellableTokensInWei = toBigNumber(oneTokenInWei).times(maximumSellableTokens).div(`1e${tokenStore.decimals}`).dp(0)

          crowdsalePageStore.setProperty('maximumSellableTokensInWei', currentMaximumSellableTokensInWei.plus(maximumSellableTokensInWei).toFixed())
        })
    })
}

export function getCurrentRate(crowdsaleContract) {
  return new Promise((resolve, reject) => {
    if (!crowdsaleContract) {
      noContractAlert()
      reject('no contract')
      return
    }

    crowdsaleContract.methods.pricingStrategy().call((err, pricingStrategyAddr) => {
      if (err) {
        console.log(err)
        reject(err)
        return
      }

      console.log('pricingStrategy:', pricingStrategyAddr)
      contractStore.setContractProperty('pricingStrategy', 'addr', pricingStrategyAddr)

      if (!pricingStrategyAddr || pricingStrategyAddr === "0x") {
        reject('no pricingStrategy address')
        return
      }

      getPricingStrategyData()
        .then(resolve)
        .catch(reject)
    });
  }
)}

export function getCrowdsaleData (crowdsaleContract) {
  return new Promise((resolve, reject) => {
    if (!crowdsaleContract) {
      noContractAlert()
      reject('no contract')
      return
    }

    console.log(crowdsaleContract)

    let propsCount = 0
    let cbCount = 0

    propsCount++
    crowdsaleContract.methods.token().call((err, tokenAddr) => {
      cbCount++

      if (err) {
        return console.log(err)
      }

      console.log('token:', tokenAddr)
      contractStore.setContractProperty('token', 'addr', tokenAddr)

      // if (propsCount === cbCount) {
      //   resolve()
      //   return
      // }

      if (!tokenAddr || tokenAddr === '0x') {
        return console.log('no token address available:', tokenAddr)
      }

      propsCount++
      getTokenData()
        .then(() => {
          if (!crowdsaleContract.methods.pricingStrategy) {
            reject('no pricingStrategy method')
            return
          }

          propsCount++
          crowdsaleContract.methods.pricingStrategy().call((err, pricingStrategyAddr) => {
            cbCount++

            if (err) {
              reject(err)
              return console.log(err)
            }

            console.log('pricingStrategy:', pricingStrategyAddr)
            contractStore.setContractProperty('pricingStrategy', 'addr', pricingStrategyAddr)

            if (propsCount === cbCount) {
              resolve()
              return
            }

            if (!pricingStrategyAddr || pricingStrategyAddr === '0x') {
              reject('no pricingStrategyAddr')
              return
            }

            getPricingStrategyData()
              .then(() => {
                cbCount++
                if (propsCount === cbCount) {
                  resolve()
                }
              })
              .catch(reject)
          })
        })
        .catch(reject)
    })
  })
}

function getTokenData () {
  return new Promise((resolve, reject) => {
    const { web3 } = web3Store

    if (!web3) {
      resolve('no MetaMask')
      return
    }

    web3.eth.getAccounts().then(accounts => {
      if (accounts.length === 0) {
        resolve('no accounts')
        return
      }

      let propsCount = 0
      let cbCount = 0
      let tokenObj = toJS(contractStore.token)
      console.log('tokenObj', tokenObj)

      attachToContract(tokenObj.abi, tokenObj.addr)
        .then(tokenContract => {
          console.log('attach to token contract')

          if (!tokenContract) {
            noContractAlert()
            reject('no contract')
            return
          }

          propsCount++
          tokenContract.methods.name().call((err, name) => {
            cbCount++

            if (err) {
              return console.log(err)
            }

            console.log('token name:', name)
            tokenStore.setProperty('name', name)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          propsCount++
          tokenContract.methods.symbol().call((err, ticker) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token ticker: ' + ticker)
            tokenStore.setProperty('ticker', ticker)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          if (tokenContract.methods.balanceOf) {
            propsCount++
            tokenContract.methods.balanceOf.call(accounts[0], (err, balanceOf) => {
              cbCount++

              if (err) {
                return console.log(err)
              }

              console.log('balanceOf: ' + balanceOf)
              const tokenAmountOf = crowdsalePageStore.tokenAmountOf ? crowdsalePageStore.tokenAmountOf : 0
              crowdsalePageStore.setProperty('tokenAmountOf', tokenAmountOf + parseInt(balanceOf, 10))

              if (propsCount === cbCount) {
                resolve()
              }
            })
          }

          propsCount++
          tokenContract.methods.decimals().call((err, decimals) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token decimals:', decimals)
            tokenStore.setProperty('decimals', decimals)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          propsCount++
          tokenContract.methods.totalSupply().call((err, supply) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token supply:', supply)
            tokenStore.setProperty('supply', supply)

            if (propsCount === cbCount) {
              resolve()

            } else {
              let propsCount = 0
              let cbCount = 0

              attachToContract(contractStore.token.abi, contractStore.token.addr)
                .then(tokenContract => {
                  console.log('attach to token contract')

                  if (!tokenContract) {
                    noContractAlert()
                    reject('no contract')
                    return
                  }

                  propsCount++
                  tokenContract.methods.name().call((err, name) => {
                    cbCount++

                    if (err) {
                      return console.log(err)
                    }

                    console.log('token name:', name)
                    tokenStore.setProperty('name', name)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  propsCount++
                  tokenContract.methods.symbol().call((err, ticker) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token ticker:', ticker)
                    tokenStore.setProperty('ticker', ticker)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  if (tokenContract.methods.balanceOf) {
                    propsCount++
                    tokenContract.methods.balanceOf(accounts[0]).call((err, balanceOf) => {
                      cbCount++

                      if (err) {
                        return console.log(err)
                      }

                      console.log('balanceOf:', balanceOf)
                      const tokenAmountOf = crowdsalePageStore.tokenAmountOf ? crowdsalePageStore.tokenAmountOf : 0
                      crowdsalePageStore.setProperty('tokenAmountOf', tokenAmountOf + parseInt(balanceOf, 10))

                      if (propsCount === cbCount) {
                        resolve()
                      }
                    })
                  }

                  propsCount++
                  tokenContract.methods.decimals().call((err, decimals) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token decimals:', decimals)
                    tokenStore.setProperty('decimals', decimals)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  propsCount++
                  tokenContract.methods.totalSupply().call((err, supply) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token supply:', supply)
                    tokenStore.setProperty('supply', supply)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })
                })
                .catch(reject)
            }
          })
        })
        .catch(reject)
    })
  })
}

export function getPricingStrategyData () {
  return new Promise((resolve, reject) => {
    attachToContract(contractStore.pricingStrategy.abi, contractStore.pricingStrategy.addr)
      .then(pricingStrategyContract => {
        console.log('attach to pricing strategy contract')

        if (!pricingStrategyContract) {
          noContractAlert()
          reject('no contract')
          return
        }

        pricingStrategyContract.methods.oneTokenInWei().call((err, rate) => {
          if (err) {
            console.log(err)
          }

          console.log('pricing strategy rate:', rate)
          crowdsalePageStore.setProperty('rate', rate)
          resolve()
        })
      })
      .catch(err => {
        reject(err)
        console.log(err)
      })
  })
}

export const getContractStoreProperty = (contract, property) => {
  const text = contractStore && contractStore[contract] && contractStore[contract][property]
  return text === undefined ? '' : text
}
