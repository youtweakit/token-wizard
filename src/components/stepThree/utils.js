import moment from 'moment';

export function defaultCompanyStartDate() {
  let crowdsaleStartDate = moment().add(5, 'minutes');
  let crowdsaleStartDateFormatted = crowdsaleStartDate.format('YYYY-MM-DDTHH:mm');
  return crowdsaleStartDateFormatted;
}

export const defaultCompanyEndDate = (startDate) => {
  const crowdsaleEndDate = moment(startDate).add(4, 'days').startOf('day')
  return crowdsaleEndDate.format('YYYY-MM-DDTHH:mm')
}
