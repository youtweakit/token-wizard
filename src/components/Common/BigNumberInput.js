import React, { Component } from 'react'
import { BigNumber } from 'bignumber.js'
import { VALIDATION_TYPES } from '../../utils/constants'
import { InputField } from './InputField'

const { VALID, INVALID } = VALIDATION_TYPES

export class BigNumberInput extends Component {
  constructor (props) {
    super(props)

    this.state = {
      value: props.value || '',
      pristine: props.pristine !== undefined ? props.pristine : true,
      valid: props.valid || VALID
    }
  }

  componentWillReceiveProps (newProps) {
    const { value, pristine, valid } = newProps

    this.setState({ value, pristine, valid })
  }

  componentDidUpdate (prevProps) {
    const { acceptEmpty, acceptFloat, minDecimals, maxDecimals, min, max } = prevProps

    if (
      prevProps.acceptEmpty !== acceptEmpty ||
      prevProps.acceptFloat !== acceptFloat ||
      prevProps.minDecimals !== minDecimals ||
      prevProps.maxDecimals !== maxDecimals ||
      prevProps.min !== min ||
      prevProps.max !== max
    ) {
      // re-check validity if any of the props had changed
      this.validate(this.state.value)
    }
  }

  validate = (value) => {
    const { acceptEmpty, acceptFloat, minDecimals, maxDecimals, min, max } = this.props
    const newState = {
      pristine: false
    }

    if (isNaN(Number(value)) || isNaN(parseFloat(value))) {
      newState.value = ''
      newState.valid = acceptEmpty ? VALID : INVALID

    } else {
      const number = new BigNumber(value)
      const decimals = number.decimalPlaces()
      let isValid = true

      if (acceptFloat) {
        if (maxDecimals !== undefined) {
          isValid = decimals <= maxDecimals
        }

        if (isValid && minDecimals !== undefined) {
          isValid = decimals >= minDecimals
        }

      } else {
        isValid = !decimals
      }

      if (isValid && min !== undefined) {
        isValid = number.gte(min)
      }

      if (isValid && max !== undefined) {
        isValid = number.lte(max)
      }

      newState.value = number.toFixed()
      newState.valid = isValid ? VALID : INVALID
    }

    this.setState(newState)
    this.props.onChange(newState)
  }

  onKeyPress = e => {
    const { acceptFloat, min, max } = this.props
    const { key } = e
    const isValidNumericKey = /[0-9.+e-]/
    const isValidIntegerKey = /[0-9-]/

    if (!isValidNumericKey.test(key)) e.preventDefault()
    if (!acceptFloat && !isValidIntegerKey.test(key)) e.preventDefault()
    if (!acceptFloat && key === '-' && min >= 0 && max >= 0) e.preventDefault()
  }

  onPaste = e => {
    if (isNaN(Number(e.clipboardData.getData('text/plain')))) e.preventDefault()
  }

  onChange = e => {
    this.validate(e.target.value)
  }

  render () {
    const { value, pristine, valid } = this.state
    const { disabled, side, errorMessage, title, description } = this.props

    return (
      <InputField
        disabled={disabled}
        side={side}
        type="text"
        errorMessage={errorMessage}
        value={value}
        pristine={pristine}
        valid={valid}
        title={title}
        onKeyPress={this.onKeyPress}
        onChange={this.onChange}
        onPaste={this.onPaste}
        description={description}
      />
    )
  }
}
