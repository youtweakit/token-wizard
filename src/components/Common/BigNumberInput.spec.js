import React from 'react'
import { BigNumberInput } from './BigNumberInput'
import { BigNumber } from 'bignumber.js'
import { DESCRIPTION, TEXT_FIELDS, VALIDATION_MESSAGES, VALIDATION_TYPES } from '../../utils/constants'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount } from 'enzyme'

configure({ adapter: new Adapter() })

describe('BigNumberInput', () => {
  const { INVALID, VALID } = VALIDATION_TYPES

  const INPUT_EVENT = {
    KEYPRESS: 'keypress',
    CHANGE: 'change',
    PASTE: 'paste'
  }

  let changeMock
  let keypressMock
  let pasteMock
  let bigNumberInputComponent
  let wrapperMemo, wrapper
  let inputMemo, input

  beforeEach(() => {
    changeMock = { target: { value: '' } }

    keypressMock = { key: '1', preventDefault: jest.fn() }

    pasteMock = {
      preventDefault: jest.fn(),
      clipboardData: {
        getData: () => 'e123e123'
      }
    }

    bigNumberInputComponent = {
      side: 'left',
      title: TEXT_FIELDS.RATE,
      description: DESCRIPTION.RATE,
      errorMessage: VALIDATION_MESSAGES.RATE,
      onChange: jest.fn()
    }

    wrapperMemo = undefined
    wrapper = () => wrapperMemo || (wrapperMemo = mount(React.createElement(BigNumberInput, bigNumberInputComponent)))

    inputMemo = undefined
    input = () => inputMemo || (inputMemo = wrapper().find('input').at(0))
  })

  it('Should render the component', () => {
    bigNumberInputComponent.min = 1
    bigNumberInputComponent.max = 1e18
    bigNumberInputComponent.acceptEmpty = false
    bigNumberInputComponent.acceptFloat = true
    bigNumberInputComponent.minDecimals = 0
    bigNumberInputComponent.maxDecimals = 4
    bigNumberInputComponent.pristine = true

    const reactElement = React.createElement(BigNumberInput, bigNumberInputComponent)
    const element = renderer.create(reactElement)

    expect(element.toJSON()).toMatchSnapshot()
  })

  describe('paste event', () => {
    [
      { value: '123', expected: 0 },
      { value: '0', expected: 0 },
      { value: '1e10', expected: 0 },
      { value: '1e+10', expected: 0 },
      { value: '1e-10', expected: 0 },
      { value: '1.23', expected: 0 },
      { value: '.123', expected: 0 },
      { value: '-123', expected: 0 },
      { value: '123e', expected: 1 },
      { value: '123e123123e12', expected: 1 },
      { value: '12345678901234567890abcd123', expected: 1 },
      { value: '123abc123', expected: 1 },
    ].forEach(testCase => {
      const action = testCase.expected ? 'fail' : 'pass'

      it(`Should ${action} for '${testCase.value}'`, () => {
        pasteMock.clipboardData.getData = () => testCase.value
        input().simulate(INPUT_EVENT.PASTE, pasteMock)

        expect(pasteMock.preventDefault).toHaveBeenCalledTimes(testCase.expected)
      })
    })
  })

  describe('key press event', () => {
    [
      {
        value: '+',
        expected: 1,
        props: { min: 0, max: 1e18 }
      },
      {
        value: '.',
        expected: 1,
        props: { min: 0, max: 1e18 }
      },
      {
        value: 'e',
        expected: 1,
        props: { min: 0, max: 1e18 }
      },
      {
        value: '-',
        expected: 1,
        props: { min: 0, max: 1e18 }
      },
      {
        value: '1',
        expected: 0,
        props: { min: 0, max: 1e18 }
      },
      {
        value: '-',
        expected: 0,
        props: { min: -10, max: 1e18 }
      },
      {
        value: '-',
        expected: 0,
        props: { max: -5 }
      },
      {
        value: '1',
        expected: 0,
        props: { min: 0, max: 1e18 }
      },
      {
        value: '.',
        expected: 0,
        props: { acceptFloat: true }
      },
      {
        value: '-',
        expected: 0,
        props: { acceptFloat: true }
      },
      {
        value: '+',
        expected: 0,
        props: { acceptFloat: true }
      },
      {
        value: 'e',
        expected: 0,
        props: { acceptFloat: true }
      },
      {
        value: 'j',
        expected: 1,
        props: { acceptFloat: true }
      },
    ].forEach(testCase => {
      const action = testCase.expected ? 'pass' : 'fail'

      it(`Should ${action} for '${testCase.value}'`, () => {
        bigNumberInputComponent.acceptFloat = testCase.props.acceptFloat
        bigNumberInputComponent.minDecimals = testCase.props.minDecimals
        bigNumberInputComponent.maxDecimals = testCase.props.maxDecimals
        bigNumberInputComponent.min = testCase.props.min
        bigNumberInputComponent.max = testCase.props.max

        keypressMock.key = testCase.value
        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)

        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(testCase.expected)
      })
    })
  })

  describe('change event', () => {
    describe('special characters', () => {
      [
        { value: 'abc', expected: INVALID },
        { value: '#@', expected: INVALID },
        { value: '~', expected: INVALID },
        { value: 'e', expected: INVALID },
        { value: '123e', expected: INVALID },
        { value: '', expected: VALID },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.acceptEmpty = testCase.expected === VALID
          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)

          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: '',
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })

    describe('with float', () => {
      [
        { value: '123', expected: VALID },
        { value: '123e3', expected: VALID },
        { value: '123e-3', expected: VALID },
        { value: '.123', expected: VALID },
        { value: '0.123', expected: VALID },
        { value: '1.123', expected: VALID },
        { value: '1.12e12', expected: VALID },
        { value: '1e-18', expected: VALID },
        { value: '1e-19', expected: INVALID },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.acceptFloat = true
          bigNumberInputComponent.min = 1e-18

          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)

          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: new BigNumber(testCase.value).toFixed(),
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })

    describe('without float', () => {
      [
        { value: '123', expected: VALID },
        { value: '123e-3', expected: INVALID },
        { value: '.123', expected: INVALID },
        { value: '0.123', expected: INVALID },
        { value: '1.123', expected: INVALID },
        { value: '10000000000000000000', expected: INVALID },
        { value: '1000000000000000000', expected: VALID },
        { value: '1000000000000000001', expected: INVALID },
        { value: '999999999999999999', expected: VALID },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.max = 1e18
          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)

          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: new BigNumber(testCase.value).toFixed(),
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })

    describe('with negative', () => {
      [
        { value: '-123', expected: VALID },
        { value: '-123e3', expected: INVALID },
        { value: '-.123', expected: VALID },
        { value: '-0.123', expected: VALID },
        { value: '-1.123', expected: VALID },
        { value: '-10000000000000000000', expected: INVALID },
        { value: '-99999', expected: VALID },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.acceptFloat = true
          bigNumberInputComponent.min = -1e5
          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)

          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: new BigNumber(testCase.value).toFixed(),
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })

    describe('without negative', () => {
      [
        { value: '-1', expected: INVALID },
        { value: '0', expected: VALID },
        { value: '15', expected: VALID },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.min = 0
          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)

          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: new BigNumber(testCase.value).toFixed(),
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })

    describe('max and min decimals', () => {
      [
        {
          value: '1e-5',
          expected: VALID,
          props: { minDecimals: 0, maxDecimals: 10}
        },
        {
          value: '1e-5',
          expected: INVALID,
          props: { minDecimals: 10, maxDecimals: 10}
        },
        {
          value: '0.0000157',
          expected: VALID,
          props: { minDecimals: 0, maxDecimals: 10}
        },
        {
          value: '0.00000000000000001544',
          expected: INVALID,
          props: { minDecimals: 0, maxDecimals: 10}
        },
        {
          value: '.6546465464654654',
          expected: VALID,
          props: { minDecimals: 15, maxDecimals: 16}
        },
        {
          value: '.654657674646546778',
          expected: INVALID,
          props: { minDecimals: 15, maxDecimals: 16}
        },
      ].forEach(testCase => {
        const action = testCase.expected === VALID ? 'pass' : 'fail'

        it(`Should ${action} for '${testCase.value}'`, () => {
          bigNumberInputComponent.acceptFloat = true
          bigNumberInputComponent.maxDecimals = testCase.props.maxDecimals
          bigNumberInputComponent.minDecimals = testCase.props.minDecimals

          changeMock.target.value = testCase.value
          input().simulate(INPUT_EVENT.CHANGE, changeMock)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledTimes(1)
          expect(bigNumberInputComponent.onChange).toHaveBeenCalledWith({
            value: new BigNumber(testCase.value).toFixed(),
            pristine: false,
            valid: testCase.expected
          })
        })
      })
    })
  })
})
