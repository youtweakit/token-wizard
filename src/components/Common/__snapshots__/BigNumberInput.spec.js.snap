// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigNumberInput Should render the component 1`] = `
<div
  className="left"
>
  <label
    className="label"
  >
    Rate
  </label>
  <input
    className="input"
    disabled={undefined}
    onBlur={undefined}
    onChange={[Function]}
    onKeyPress={[Function]}
    onPaste={[Function]}
    style={undefined}
    type="text"
    value=""
  />
  <p
    className="description"
  >
    Exchange rate Ethereum to Tokens. If it's 100, then for 1 Ether you can buy 100 tokens
  </p>
  <p
    style={
      Object {
        "color": "red",
        "fontSize": "12px",
        "fontWeight": "bold",
        "height": "10px",
        "width": "100%",
      }
    }
  />
</div>
`;
